ath79: add support for D-Link DAP-1330/DAP-1365 A1

s-2 opened the pull request:
Port device support for DAP-1330 from the ar71xx target to ath79.

Additionally, images are generated for the European through-socket
case variant DAP-1365. Both devices run the same vendor firmware, the
only difference being the `DAP_SIGNATURE` field in the factory header.
The vendor's Web UI will display a model string stored in the flash.

Specifications:

 * QCA9533, 8 MiB Flash, 64 MiB RAM
 * One Ethernet Port (10/100)
 * Wall-plug style case (DAP-1365 with additional socket)
 * LED bargraph RSSI indicator

Installation:

 * Web UI: http://192.168.0.50 (or different address obtained via DHCP)
   There is no password set by default
 * Recovery Web UI: Keep reset button pressed during power-on
   until LED starts flashing red, upgrade via http://192.168.0.50
 * Some modern browsers may have problems flashing via the Web UI,
   if this occurs consider booting to recovery mode and flashing via:
   curl -F \
     files=@openwrt-ath79-generic-dlink_dap-1330-a1-squashfs-factory.bin \
     http://192.168.0.50/cgi/index

The device will use the same MAC address for both wired and wireless
interfaces, however it is stored at two different locations in the flash.

Factory image generation is based on `mkdapimg2` from ar71xx, which might
already be introduced to ath79 by DCH-G020 from PR #2998.
Also these devices share the same MAC layout, thus the corresponding
base-files entries (in `02_network` and `10_fix_wifi_mac`) could be merged.

LED naming is a little awkward, but compatible to ar71xx.

Apparently using a .dtsi is currently the best method to include
identical devices which only differ in the model name. Makefile entries
are also hierarchical to keep `SUPPORTED_DEVICES` out of DAP-1365 metadata.

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-1330-a1.dts:
```diff
@@ -0,0 +1,9 @@
+
+/ {
+	compatible = "dlink,dap-1330-a1", "qca,qca9533";
+	model = "D-Link DAP-1330-A1";
```
No hyphen before A1?

s-2 replied:
It's amazing how I could read 10x through the whole diff, and still there's always these kinds of things left...

fragtion replied:
What's even more amazing is how @adrianschmutzler picks these things up without error. What a wizard! 😂

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-1365-a1.dts:
```diff
@@ -0,0 +1,9 @@
+
+/ {
+	compatible = "dlink,dap-1365-a1", "qca,qca9533";
+	model = "D-Link DAP-1365-A1";
```
No hyphen before A1?

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-13xx.dtsi:
```diff
@@ -0,0 +1,127 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_power: power_green {
```
`led_power_green: power_green {`

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -101,6 +101,14 @@ engenius,ecb1750)
+	ucidef_set_led_rssi "wifi-low" "wifi-low" "d-link:red:wifi" "wlan0" "1" "29"
+	ucidef_set_led_rssi "wifi-medium" "wifi-medium" "d-link:green:wifi" "wlan0" "30" "100"
+	ucidef_set_led_rssi "wifi-high" "wifi-high" "d-link:green:signal1" "wlan0" "50" "100"
+	ucidef_set_led_rssi "wifi-max" "wifi-max" "d-link:green:signal2" "wlan0" "70" "100"
```
I see that you copied this from ar71xx. But if you ask me, this is nonsense:
- 2000000 should be 200000 (and is it really 2 for the last parameter)
- wifi_red/wifi_green look to me like the should be status indicators
- wifi-low parameters don't make sense

Please research what behavior these LEDs should have based on vendor behavior.

s-2 replied:
The RSSI graph consists of three fields, only the lower one being composed of two LEDs red and green. So I would assume this would change from green to red when the signal is very low (which I honestly could not reproduce even when using tin foil, is there any way the RSSI value could be mocked?). I'll check the rssileds package again and dig into the vendor behaviour.

adrianschmutzler replied:
Ah, I got it now. So, the arguments for the ucidef_set_led_rssi lines are actually correct that way, as the red led will turn off after 29 % and the "medium" one will only start to light at 30 %. So, the only thing to change would be the names, and check the rssimon parameters, of course.

adrianschmutzler replied:
Despite, I wouldn't use too much time on testing that. I'd just rename the LEDs and use the standard parameters for rssimon ...

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -132,6 +132,11 @@ ath79_setup_interfaces()
+	dlink,dap-1330-a1|\
+	dlink,dap-1365-a1)
+		ucidef_add_switch "switch0" \
+			"0@eth0" "1:lan:1"
```
Why not just
ucidef_set_interface_lan "eth0"

s-2 replied:
I guess the `Switch` menu item appearing on LuCI (for port status and VLAN configuration) is not a legitimate argument here, so it might as well be just LAN on all eth0.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -448,6 +459,32 @@ define Device/devolo_magic-2-wifi
+define Device/dlink_dap-13xx
+  SOC := qca9533
+  DEVICE_VENDOR := D-Link
+  DEVICE_VARIANT := A1
```
I'd prefer the variant at the device nodes.

s-2 replied:
Also thought about that, and agree to prefer it over reducing redundancy here.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -448,6 +459,32 @@ define Device/devolo_magic-2-wifi
+  DEVICE_VARIANT := A1
+  DEVICE_PACKAGES += rssileds
+  IMAGES += factory.bin
+  IMAGE_SIZE := 7936k
```
move image_size, so IMAGES is next to the corresponding definition.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-13xx.dtsi:
```diff
@@ -0,0 +1,127 @@
+
+/ {
+	aliases {
+		led-boot = &led_power_green;
```
Based on the manual the red color indicates booting.

s-2 replied:
During u-boot execution, the status LED is red, then goes off for about two seconds before it starts the typical (green) OpenWRT boot flashing pattern. This is what I am used to from other OpenWRT devices, though honestly I'm not even exactly sure what the led-boot alias is for; also, would we need gpio-hog or something to get rid of the short off period?

adrianschmutzler replied:
what's written in the aliases section is executed here: https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh

So, this will be OpenWrt's boot blinking pattern. To be honest, I don't care so much about whether this LED blinks red or green then, but the "proper" way would be to do it the same way as the vendor (if that's possible). As a bonus, you will then have different colors for boot and failsafe, for example ...

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-13xx.dtsi:
```diff
@@ -0,0 +1,127 @@
+		};
+
+		wifi_red {
+			label = "d-link:red:wifi";
```
d-link:red:rssilow

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-13xx.dtsi:
```diff
@@ -0,0 +1,127 @@
+		};
+
+		wifi_green {
+			label = "d-link:green:wifi";
```
d-link:green:rssimedium

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-13xx.dtsi:
```diff
@@ -0,0 +1,127 @@
+		};
+
+		signal1 {
+			label = "d-link:green:signal1";
```
d-link:green:rssihigh

adrianschmutzler commented on target/linux/ath79/dts/qca9533_dlink_dap-13xx.dtsi:
```diff
@@ -0,0 +1,127 @@
+		};
+
+		signal2 {
+			label = "d-link:green:signal2";
```
d-link:green:rssimax

If you don't like that, you could also use the following scheme:
rssilow
rssimediumlow
rssimediumhigh
rssihigh

The latter is used more frequently

s-2 replied:
This looks like a good idea. Do we need an additional LED migration script in base-files then?
However, though flashing from ar71xx worked (due to `SUPPORTED_DEVICES` entry) I did not manage to connect to the device when choosing to keep settings, not sure if this is a generic problem or could be fixed within the new image.

adrianschmutzler replied:
LED migration would be set up here: https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration

network setup is always an issue when updating from ar71xx to ath79, that's why we recommend to start with clean config. I haven't actually looked at the config for ar71xx yet, so I can't tell any details. However, there is no entry in 02_network there, so the device actually sets up an eth1 while it doesn't have one:

		ucidef_set_interfaces_lan_wan "eth0" "eth1"

adrianschmutzler commented:
Found the manual: https://cc.cnetcontent.com/inlinecontent/mediaserver/dli/4ae/e29/4aee2994162f49aeb27db3f11acd6c77/original.pdf

s-2 commented:
> Found the manual: 
Since they don't use the 'red only' state, we'd just stick with the 4 RSSI levels as they do in the vendor's firmware.

s-2 commented:
This `rssileds` package is driving me nuts...
At first it seemed like overlapping rules for `rssilow` and `rssimediumlow` were not supported, but it also works for the others, i.e. the lower green ones remain on when the signal gets better. The only reproducible way of using the red led is setting it to `"1" "100"`, i.e. having it remain on even with better signal, which is not the desired behaviour.

I also tried setting the brightness offset, setting `qmin` down to `0` etc... eventually it turns out it just worked sporadically sometimes, after reboot the red led would not turn on when the signal drops below 25%. According to `rssileds.c` the default brightness of 0 should have no effect:
```
void update_leds(rule_t *rules, int q)
{
        rule_t *rule = rules;
        while (rule)
        {
                int b;
                /* offset and factore correction according to rule */
                b = ( q + rule->boffset ) * rule->bfactor;
                if ( b < 0 )
                        b=0;
                if ( b > 255 )
                        b=255;

                if ( q >= rule->minq && q <= rule->maxq )
                        set_led(rule->led, (unsigned char)b);
                else
                        set_led(rule->led, 0);

                rule = rule->next;
        }
}
```
Otherwise the led is working perfectly, writing to `/brightness` will turn it on for any value greater 0.
This probably needs further investigation outside of this PR; I will try to add some debug outputs to rssileds.

For now, I am leaving the configuration consistent with other devices (refresh, threshold (hysteresis) and led thresholds (signal quality), as well as the manufacturer's behaviour (use amber isntead of red for the lowest range).

Regarding the boot blinking pattern, I tried using the red LED, but it just felt wrong; considering the boot process itself is always something that could be considered positive, and red colour being reserved for indicating faults or special states, it is now used for failsafe and upgrade modes only.

s-2 commented:
Now that I've added uart outputs to the rssidleds update-method and flashed another device, it's obviously working flawlessly now... 
No idea what's different between the devices, but I found the DAP-1330 factory image is broken - actually both images are being generated with the DAP1365 signature, I'll check tomorrow... (getting tired and a little paranoid - apparently Vodafone is now blocking my association requests to their Hotspot network, should probably change wifi MAC, or they have actually implemented some band steering / load balancing and reject devices that deliberately connect to the farthest AP they can find...)

s-2 commented:
Turns out `DEVICE_VARS` is still required in ath79, but was just moved to the top of the Makefile since ar71xx.
Factory images are building correctly now even with multiple profiles selected.

I wonder if it needs to be updated in #2998 as well, although there would be no problems until another device (e.g. this PR which fixed it) would be merged. However, I'm still thinking about changing the LED upgrade/failsafe behaviour in DCH-G020 to match DAP-1330.

adrianschmutzler commented:
To be save, you should select "Multiple devices" and "rootfs per target" ...

s-2 committed:
ath79: add support for D-Link DAP-1330/DAP-1365 A1

Port device support for DAP-1330 from the ar71xx target to ath79.

Additionally, images are generated for the European through-socket
case variant DAP-1365. Both devices run the same vendor firmware, the
only difference being the DAP_SIGNATURE field in the factory header.
The vendor's Web UI will display a model string stored in the flash.

Specifications:

 * QCA9533, 8 MiB Flash, 64 MiB RAM
 * One Ethernet Port (10/100)
 * Wall-plug style case (DAP-1365 with additional socket)
 * LED bargraph RSSI indicator

Installation:

 * Web UI: http://192.168.0.50 (or different address obtained via DHCP)
   There is no password set by default
 * Recovery Web UI: Keep reset button pressed during power-on
   until LED starts flashing red, upgrade via http://192.168.0.50
 * Some modern browsers may have problems flashing via the Web UI,
   if this occurs consider booting to recovery mode and flashing via:
   curl -F \
     files=@openwrt-ath79-generic-dlink_dap-1330-a1-squashfs-factory.bin \
     http://192.168.0.50/cgi/index

The device will use the same MAC address for both wired and wireless
interfaces, however it is stored at two different locations in the flash.

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 commented:
Rebased to resolve merge conflict in `01_leds` due to DIR-842 removal from #3027.

ynezz commented:
Looks good to me, please rebase.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -136,6 +136,10 @@ ath79_setup_interfaces()
+	dlink,dap-1330-a1|\
+	dlink,dap-1365-a1)
+		ucidef_set_interface_lan "eth0"
+		;;
```
This could be merged with the big block at the top of the case ...

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -359,6 +363,11 @@ ath79_setup_macs()
+	dlink,dap-1365-a1)
+		lan_mac=$(mtd_get_mac_text "mp" 0x1)
+		label_mac=$lan_mac
+		;;
```
@ynezz If you merge both dlink PRs from this author, I think this node can be merged for them.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:
```diff
@@ -17,6 +17,10 @@ case "$board" in
+	dlink,dap-1330-a1|\
+	dlink,dap-1365-a1)
+		mtd_get_mac_text "mp" 0x13 > /sys${DEVPATH}/macaddress
+		;;
```
@ynezz Same here, can be merged with dlink,dch-g020-a1)

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration:
```diff
@@ -7,6 +7,11 @@ board=$(board_name)
+dlink,dap-1330-a1)
+	migrate_leds ":red:power=:red:status" \
+		":red:wifi=:red:rssilow" ":green:wifi=:green:rssimediumlow" \
+		":green:signal1=:green:rssimediumhigh" ":green:signal2=:green:rssihigh"
```
Nitpick, but this might be nicer with one line per replacement:
```
	migrate_leds ":red:power=:red:status" \
		":red:wifi=:red:rssilow" \
		":green:wifi=:green:rssimediumlow" \
		":green:signal1=:green:rssimediumhigh" \
		":green:signal2=:green:rssihigh"
```

adrianschmutzler commented:
This might be made easier by adding this rebased on top of dlink,dch-g020-a1 PR, but I don't know whether @ynezz would prefer this.

ynezz commented:
So please rebase, thanks.

s-2 commented:
Rebased, pushed.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
